SSA/Projects/InstCombine/HackersDelight.lean

```diff
@@ -304,6 +307,89 @@ theorem sle_iff_adc_two_pow_sub_neg_add_two_pow_sub :
     (x ≤ₛ 0) ↔ (BitVec.carry w (2 ^ (w - 1)) (-(x + 2 ^ (w - 1)))) false := by
   sorry
 
+theorem add_iff_not_ult :
+    AdditionNoOverflows? x y ↔ ~~~ x <ᵤ y := by
+  sorry
+
+theorem add_iff_add_ult :
+    AdditionNoOverflows? x y ↔ x + y <ᵤ x := by
+  sorry
+
+theorem add_add_iff_not_ule :
+    AdditionNoOverflows? x (y + 1) ↔ ~~~ x ≤ᵤ y := by
+  sorry
+
+theorem add_add_iff_add_add_ule :
+    AdditionNoOverflows? x (y + 1) ↔ x + y + 1 ≤ᵤ x := by
+  sorry
+
+theorem add_not_add_iff_ult :
+    AdditionNoOverflows? x (~~~ y + 1) ↔ x <ᵤ y := by
+  sorry
+
+theorem add_not_add_iff_ult_sub :
+    AdditionNoOverflows? x (~~~ y + 1) ↔ x <ᵤ x - y := by
+  sorry
+
+theorem add_not_iff_ult :
+    AdditionNoOverflows? x (~~~ y) ↔ x ≤ᵤ y := by
+  sorry
+
+theorem add_not_iff_ule_sub_sub :
+    AdditionNoOverflows? x (~~~ y) ↔ x ≤ᵤ x - y - 1 := by
+  sorry
+
+def UnsignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ w
+def SignedMultiplicationOverflows? (x y : BitVec w) : Prop := x.toNat * y.toNat > 2 ^ (w - 1)
+
+def Last32Bits (x : BitVec 64) : BitVec 32 := BitVec.ofNat 32 x.toNat
+def First32Bits (x : BitVec 64) : BitVec 32 := BitVec.ofNat 32 (x >>> 32).toNat
+
+theorem unsigned_mul_overflow_iff_mul_neq_zero {x y : BitVec 64} :
+    UnsignedMultiplicationOverflows? x y ↔ First32Bits (x * y) ≠ BitVec.ofNat 32 0 := by
+  sorry
+
+theorem signed_mul_overflow_iff_mul_neq_mul_RightShift {x y : BitVec 64} :
+    SignedMultiplicationOverflows? x y ↔ First32Bits (x * y) ≠ Last32Bits (x * y) >>> 31 := by
+  sorry
+
+theorem mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (x * y) / z ≠ x → UnsignedMultiplicationOverflows? x y := by
+  sorry
+
+theorem lt_and_eq_neg_pow_two_or_mul_div_neq_imp_unsigned_mul_overflow (h : y.toNat ≠ 0) :
+    (y < 0) && (x.toInt = - 2 ^ 31) || ((x * y) / z ≠ x) → SignedMultiplicationOverflows? x y := by
+  sorry
+
+def NumberOfLeadingZeros {w : Nat} (x : BitVec w) : Nat :=
```

Comment:
This theorem name should start with a lowercase letter

https://leanprover-community.github.io/contribute/naming.html

Reply:
Via rule 3 and then rule 4.